### PR TITLE
CommercioDOCS improvements

### PR DESCRIPTION
**Pull request by RiccardoM**
### Changes
- Split the various document related types into many files to better handle them
- Changed the way the documents are stored inside the chain state: now they are all stored using their ids, without replicating them for the sender and the recipient. Those two will have a list of ids instead of documents now
- Removed any internal import inside the tests using the aliases as descried inside #26
- Added the possibility of sending one document to multiple recipients using one single transaction
- Added the option to specify an encryption key for each recipient and to specify which data has been encrypted

### CODE REVIEW

File: x/docs/internal/keeper/keeper_test.go [hunk 1]
```diff
@@ -190,125 +190,173 @@ func TestKeeper_ShareDocument_EmptyLists(t *testing.T) {
...
+
+	assert.Equal(t, 2, len(receivedDocs))
+	assert.Contains(t, receivedDocs, TestingDocument.Uuid)
 }
 
 func TestKeeper_GetUserReceivedDocuments_EmptyList(t *testing.T) {
 	_, ctx, k := SetupTestInput()
 	store := ctx.KVStore(k.StoreKey)
-	store.Delete([]byte(types.ReceivedDocumentsPrefix + TestingRecipient.String()))
+	store.Delete(k.getReceivedDocumentsStoreKey(TestingRecipient))
```
Comment (bragaz) [inline x/docs/internal/keeper/keeper_test.go]: you shouldn't need this if you keep redeclaring store in every test with new `ctx` and `k`

File: x/docs/internal/keeper/keeper_test.go [hunk 2]
```diff
@@ -190,125 +190,173 @@ func TestKeeper_ShareDocument_EmptyLists(t *testing.T) {
...
 	assert.Equal(t, 1, len(receivedDocs))
 	assert.Equal(t, documents, receivedDocs)
 }
 
 func TestKeeper_GetUserSentDocuments_EmptyList(t *testing.T) {
 	_, ctx, k := SetupTestInput()
+
 	store := ctx.KVStore(k.StoreKey)
-	store.Delete([]byte(types.SentDocumentsPrefix + TestingSender.String()))
+	store.Delete(k.getSentDocumentsStoreKey(TestingSender))
```
Comment (bragaz) [inline x/docs/internal/keeper/keeper_test.go]: you shouldn't need this if you keep initializing store in every test with new `ctx` and ´k`

File: x/docs/internal/keeper/keeper_test.go [hunk 3]
```diff
@@ -440,16 +488,16 @@ func TestKeeper_GetUsersSet_EmptySet(t *testing.T) {
 func TestKeeper_SetUserDocuments(t *testing.T) {
 	cdc, ctx, k := SetupTestInput()
 	store := ctx.KVStore(k.StoreKey)
-	store.Delete(k.getSentDocumentsStoreKey(TestingDocument.Sender))
-	store.Delete(k.getReceivedDocumentsStoreKey(TestingDocument.Recipient))
+	store.Delete(k.getSentDocumentsStoreKey(TestingSender))
+	store.Delete(k.getReceivedDocumentsStoreKey(TestingRecipient))
```
Comment (bragaz) [inline x/docs/internal/keeper/keeper_test.go]: you shouldn't need this if you keep initializing store in every test with new `ctx` and `k`

File: x/docs/internal/keeper/querier_test.go [hunk 4]
```diff
@@ -21,9 +21,9 @@ func Test_queryGetReceivedDocuments_EmptyList(t *testing.T) {
 	var querier = NewQuerier(k)
 	//Setup the store
 	store := ctx.KVStore(k.StoreKey)
-	store.Delete(k.getReceivedDocumentsStoreKey(TestingDocument.Recipient))
+	store.Delete(k.getReceivedDocumentsStoreKey(TestingRecipient))
```
Comment (bragaz) [inline x/docs/internal/keeper/querier_test.go]: you shouldn't need this if you keep initializing store in every test with new `ctx` and `k`

File: x/docs/internal/keeper/querier_test.go [hunk 5]
```diff
@@ -59,9 +59,9 @@ func Test_queryGetSentDocuments_EmptyList(t *testing.T) {
 	var querier = NewQuerier(k)
 	//Setup the store
 	store := ctx.KVStore(k.StoreKey)
-	store.Delete(k.getSentDocumentsStoreKey(TestingDocument.Sender))
+	store.Delete(k.getSentDocumentsStoreKey(TestingSender))
```
Comment (bragaz) [inline x/docs/internal/keeper/querier_test.go]: you shouldn't need this if you keep initializing store in every test with new `ctx` and `k`

File: x/docs/internal/types/document_enryption_key.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	if len(key.Value) == 0 {
+		return errors.New("encryption key value cannot be null")
+	}
+
+	if len(key.Encoding) == 0 {
+		return errors.New("encryption key encoding cannot be null")
+	}
+
+	encoding := strings.ToLower(key.Encoding)
+	if encoding != "base64" && encoding != "hex" {
```
Comment (bragaz) [inline x/docs/internal/types/document_enryption_key.go]: i think you would write  `if encoding != "base64" || encoding != "hex"`

File: x/docs/internal/types/document_metadata.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	return metadata.ContentUri == other.ContentUri &&
+		metadata.Proof == other.Proof &&
+		metadata.Schema.Equals(*other.Schema)
+}
+
+// Validate tries to validate all the data contained inside the given
+// DocumentMetadata and returns an error if something is wrong
+// TODO: Test this
+func (metadata DocumentMetadata) Validate() error {
+	if len(metadata.ContentUri) == 0 {
```
Comment (bragaz) [inline x/docs/internal/types/document_metadata.go]: does `len` return `0` even if strings are  filled with blank spaces?? we should check this and eventually, changes these checks

File: x/docs/internal/types/document_metadata_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+	actual := invalidDocumentMetadata.Validate()
+	assert.NotNil(t, actual)
+}
+
+func TestValidateDocMetadata_emptySchemaUri(t *testing.T) {
+	invalidDocumentMetadata := DocumentMetadata{
+		ContentUri: "http://www.contentUri.com",
+		Schema: &DocumentMetadataSchema{
+			Uri:     "",
```
Comment (bragaz) [inline x/docs/internal/types/document_metadata_test.go]: we should test even if `Uri` is a string like this one `"     "`

File: x/docs/internal/types/msgs.go [hunk 9]
```diff
@@ -34,87 +40,56 @@ func (msg MsgShareDocument) Route() string { return ModuleName }
...
+// TODO: Test more
+func (msg MsgShareDocument) ValidateBasic() sdk.Error {
+	if msg.Sender.Empty() {
+		return sdk.ErrInvalidAddress(msg.Sender.String())
 	}
 
-	if (docMetadata.Schema == nil) && len(docMetadata.SchemaType) == 0 {
-		return sdk.ErrUnknownRequest("Either schema or schema_type must be defined")
+	if msg.Recipients.Empty() {
+		return sdk.ErrUnknownRequest("Recipients cannot be empty")
```
Comment (bragaz) [inline x/docs/internal/types/msgs.go]: This error probably should be an `sdk.ErrInvalidAddress` like the one above

File: x/docs/internal/types/document_metadata_test.go (hunk 8 above)
Comment (RiccardoM) [reply x/docs/internal/types/document_metadata_test.go]: Done

File: x/docs/internal/types/document_metadata.go (hunk 7 above)
Comment (RiccardoM) [reply x/docs/internal/types/document_metadata.go]: Done, fixed all the checks

File: x/docs/internal/types/msgs.go (hunk 9 above)
Comment (RiccardoM) [reply x/docs/internal/types/msgs.go]: Good point, changed this

File: x/id/internal/keeper/keeper_test.go [hunk 10]
```diff
@@ -54,7 +54,7 @@ func TestKeeper_GetIdentities(t *testing.T) {
 	store := ctx.KVStore(k.StoreKey)
 	iterator := sdk.KVStorePrefixIterator(store, []byte(types.IdentitiesStorePrefix))
 	for ; iterator.Valid(); iterator.Next() {
```
Comment (bragaz) [inline x/id/internal/keeper/keeper_test.go]: This cycle is useless now

Comment (RiccardoM) [reply x/id/internal/keeper/keeper_test.go]: Removed

**Pull request merged**